src/mpid/ch4/netmod/ofi/ofi_am_impl.c

```diff
@@ -66,6 +67,92 @@ static int issue_deferred_am_isend_eager(MPIDI_OFI_deferred_am_isend_req_t * dre
     goto fn_exit;
 }
 
+static int issue_deferred_am_isend_pipeline(MPIDI_OFI_deferred_am_isend_req_t * dreq)
+{
+    int mpi_errno = MPI_SUCCESS;
+    char *send_buf;
+    MPI_Aint packed_size;
+    MPI_Aint dt_true_lb, send_size;
+    MPIR_Datatype *dt_ptr;
+    MPIDI_OFI_am_send_pipeline_request_t *send_req;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_ISSUE_DEFERRED_AM_ISEND_PIPELINE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_ISSUE_DEFERRED_AM_ISEND_PIPELINE);
+
+    send_size = MPIDI_OFI_DEFAULT_SHORT_SEND_SIZE - dreq->am_hdr_sz - sizeof(MPIDI_OFI_am_header_t);
+    send_size = MPL_MIN(send_size, MPIDIG_send_get_data_sz_left(dreq->sreq));
+
+    if (dreq->need_packing) {
+        /* FIXME: currently we always do packing, also for high density types. However,
+         * we should not do packing unless needed. Also, for large low-density types
+         * we should not allocate the entire buffer and do the packing at once. */
+        /* TODO: (1) Skip packing for high-density datatypes; */
+        mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
+                                                       (void **) &send_buf);
+        if (send_buf == NULL) {
+            /* no buffer available, suppress error and exit */
+            mpi_errno = MPI_SUCCESS;
+            goto fn_exit;
+        }
+        MPIR_ERR_CHECK(mpi_errno);
+        mpi_errno = MPIR_Typerep_pack(dreq->buf, dreq->count, dreq->datatype,
+                                      MPIDIG_send_get_offset(dreq->sreq), send_buf, send_size,
+                                      &packed_size);
+        MPIR_ERR_CHECK(mpi_errno);
+        send_size = packed_size;
+
+        MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.am_hdr_buf_pool, (void **) &send_req);
+        MPIR_Assert(send_req);
+        send_req->sreq = dreq->sreq;
+        if (dreq->am_hdr_sz > 0) {
+            send_req->am_hdr_sz = dreq->am_hdr_sz;
+            send_req->am_hdr = (void *) &send_req->am_hdr_buf[0];
+            MPIR_Memcpy(send_req->am_hdr, dreq->am_hdr, dreq->am_hdr_sz);
```

Comment:
Do we need this `memcpy` since `dreq->am_hdr` is already in the request?

Reply:
Yes. We should be use the one in deferred request because it could be freed before the send is completed.